feat: Add AST serialization and exporting support

YaCodesDevelopment opened the pull request:
Added `--save_ast` flag. When passed, this should also output the `ast.json` file while compiling. The file includes an array of ASTs for all the files compiled.
This allows the use of that `ast.json` for further analysis of code in third-party libraries, unlocking the ability to create linters and other helper tools.

YaCodesDevelopment committed:
feat: Add AST serialization and exporting support

Added --save_ast flag

Arvolear commented on parser/src/lib.rs:
```diff
@@ -109,6 +116,11 @@ pub fn run_parser(
     }
 
+    if save_ast {
+        let ast_list= serde_json::to_string(&ast_list).unwrap();
```
```suggestion
        let ast_list = serde_json::to_string(&ast_list).unwrap();
```

YaCodesDevelopment committed:
fix: add argument parser function for --save_ast

YaCodesDevelopment committed:
feat: Add path input for ast file

Now it's possible to specify default file path/name for the serialized AST file output

YaCodesDevelopment committed:
feat: Add --dry-run to skip compilation

When --dry-run is specified, it will only run check and output serialized AST, but not compile

YaCodesDevelopment committed:
fix: save_ast always true, due to default value

Removed default value.
Hint improved.

Arvolear commented on circom/src/input_user.rs:
```diff
@@ -344,6 +351,41 @@ mod input_processing {
+    }
+
+    pub fn get_ast_path(matches: &ArgMatches) -> PathBuf {
+        
+        match matches.is_present("save_ast"){
+            true => 
+               {
+                   let path = matches.value_of("save_ast").unwrap();
+                   if path != "" {
+                        let path = Path::new(path);
+                        if let Some(parent) = path.parent() {
+                            std::fs::create_dir_all(parent).unwrap();
+                        }
+                        if path.is_dir() {
+                            PathBuf::from(path.join("ast.json").to_str().unwrap())
+                        }
+                        else {
+                            PathBuf::from(path.to_str().unwrap())
+                        }
+                    }
+                    else {
+                        PathBuf::from("ast.json")
+                    }
+               }
+               
+            false => PathBuf::from("ast.json"),
+        }
```
```suggestion
        match matches.is_present("save_ast"){
            true => 
               {
                   let path = matches.value_of("save_ast").unwrap();
                   if path != "" {
                        let path = Path::new(path);
                        if let Some(parent) = path.parent() {
                            std::fs::create_dir_all(parent).unwrap();
                        }
                        
                        if path.is_dir() {
                            PathBuf::from(path.join("ast.json").to_str().unwrap())
                        } else {
                            PathBuf::from(path.to_str().unwrap())
                        }
                    } else {
                        PathBuf::from("ast.json")
                    }
               }
               
            false => PathBuf::from("ast.json"),
        }
```

YaCodesDevelopment committed:
style: Improve code style

Co-authored-by: Artem Chystiakov <47551140+Arvolear@users.noreply.github.com>

YaCodesDevelopment committed:
fix: save_ast required a value, change path logic

Fix save_ast required a value, despite being optional.
Change path handling logic in save_ast to output into the given output path

Arvolear reviewed (approved):
LGTM! (let's wait for the second approval)

velykodnyi commented on circom/src/input_user.rs:
```diff
@@ -344,6 +351,41 @@ mod input_processing {
+        matches.occurrences_of("save_ast") > 0
+    }
+
+    pub fn get_ast_path(matches: &ArgMatches) -> PathBuf {
```
Return Result and handle errors correctly(like you have in get_prime)

velykodnyi commented on parser/src/lib.rs:
```diff
@@ -109,6 +117,11 @@ pub fn run_parser(
     }
 
+    if save_ast {
+        let ast_list = serde_json::to_string(&ast_list).unwrap();
```
Panics again

YaCodesDevelopment committed:
fix: Remove unwraps to avoid panicking

velykodnyi reviewed (approved):
LGTM
